XXX_Feature/account/models_XXX 8d

HerdaSinama opened the pull request.

HerdaSinama committed:
feat: добавлена модель User

Кайф вообще круто!
- реализовано добавление картинки (и ее валидация)

HerdaSinama committed:
feat: model Team has been added to apps/account

-added model fields (Team) according to task

HerdaSinama committed:
feat: separation from the Team model member1/member2

- added the TeamMembership class
- the Team model no longer has fields with roles (
they are located in TeamMembership)
- each user can only be in one command

HerdaSinama committed:
chore: fixed a typo in the application name and added a file services.py

- rename account app (was acoont)
- add services.py for models

HerdaSinama committed:
Add simple methods for Team model

new Team method:
- add member in team
- remove member in team
- set team status active/disbanded
- get members

HerdaSinama committed:
Fix incorrect imports and misspell

- incorrect import QS
- misspell in class name in apps.py

HerdaSinama committed:
Deleted base class

- remove DefaultModel

HerdaSinama committed:
Fix functions for creator in Team model, Init serializers.py

- add func: is_creator which checks user role
- func remove_member, set_status_active, set_status_disbanded:  cannot set same status, only creator has such rights

HerdaSinama committed:
Add serializers for User model

- UserSerializer: get core user data
- UserDetailSerializer: get all user data
- UserUpdateSerializer: stores mutable fields

ArtyomYapryntsev commented on server/apps/acсount/views.py:
```diff
@@ -0,0 +1,13 @@
+
+class UserListView(generics.ListAPIView):
+    serializer_class = UserSerializer
+    permission_classes = (IsAuthenticated)
```
The `permission_classes` attribute should be iterable. In your case, `permission_classes` is equal to the `IsAuthenticated` class.
Try converting it to a tuple or a list of classes.

Source: https://www.django-rest-framework.org/api-guide/generic-views/#examples

ArtyomYapryntsev commented on server/apps/acсount/models.py:
```diff
@@ -0,0 +1,200 @@
+    """Валидатор размера файла для изображений."""
+    limit = 5 * 1024 * 1024
+    if file.size > limit:
+        raise ValidationError('нюхай бебру')
```
Edit the `ValidationError` message 😃

ArtyomYapryntsev commented on server/apps/acсount/models.py:
```diff
@@ -0,0 +1,200 @@
+    gender = models.CharField(
+        max_length=6,
+        choices=GENDERS.choices,
+        default=GENDERS.MALE,
```
Why is the user `male` by default???

It would be better if u allow the `None` value and set it to the default.

ArtyomYapryntsev commented on server/apps/acсount/models.py:
```diff
@@ -0,0 +1,200 @@
+        )
+    avatar = models.ImageField(
+        verbose_name='Avatar',
+        upload_to='media/',
```
If u trying to upload some image, u will catch an error.

Before using `ImageField`, you need to define MEDIA_ROOT in the `settings` file.

Source:
https://docs.djangoproject.com/en/5.1/ref/settings/#media-root
https://docs.djangoproject.com/en/5.1/ref/models/fields/#django.db.models.FileField.storage

ArtyomYapryntsev commented on server/apps/acсount/models.py:
```diff
@@ -0,0 +1,200 @@
+        verbose_name='Команда'
+        verbose_name_plural='Команды'
+    
+    def add_member(self, user: User, reole: str = 'participant') -> None:
```
Typo in the `reole` word

ArtyomYapryntsev commented on server/apps/acсount/models.py:
```diff
@@ -0,0 +1,200 @@
+        """
+        if not TeamMembership.objects.filter(team=self, user=user).exists():
+            TeamMembership.objects.create(team=self, user=user, reole=reole)
+        raise ValidationError(
```
The `ValidationError` will be raised in both cases

ArtyomYapryntsev commented on server/apps/acсount/models.py:
```diff
@@ -0,0 +1,200 @@
+
+    def remove_member(
+            self,
+            user_to_remover: User,
```
I guess u want to enter `user_to_remove`

ArtyomYapryntsev commented on server/apps/acсount/models.py:
```diff
@@ -0,0 +1,200 @@
+        """
+        if not self.is_creator(current_user):
+            raise ValidationError('Вы не являетесь создателем команды')
+        if user_to_remover == current_user:
```
Check this before `not self.is_creator(...)`.

The `not self.is_creator(...)`  condition requires an additional db query. So, if the `remove_member` method receives an isntance of the `User` model through the `user_to_remove` and `current_user` fields, then the method will not execute a redundant db query.

HerdaSinama replied:
if we pass the queryset along with the user role, we will still send an additional request to the database. or I don't understand smth?

HerdaSinama replied:
nevermind i get it

ArtyomYapryntsev commented on server/apps/acсount/models.py:
```diff
@@ -0,0 +1,200 @@
+
+    def get_members(self) -> QuerySet[Any]:
+        """Возвращает QuerySet участников команды."""
+        return self.membership.select_related('user').all()
```
Using the `select_related` method. Very nice!!!

ArtyomYapryntsev commented on server/apps/acсount/models.py:
```diff
@@ -0,0 +1,200 @@
+class TeamMembership(models.Model):
+    class ROLECHOICES(TextChoices):
+        CREATOR = 'creator', 'Creator'
+        PARTICIPANT = 'participant', 'Participant'
```
May be replace `Participant` with `Member`?

What do u think abuot that?

ArtyomYapryntsev commented on server/apps/acсount/serializers.py:
```diff
@@ -0,0 +1,66 @@
+from .models import User, Team
+
+
+class UserSerializer(serializers.ModelSerializer):
+    full_name = serializers.SerializerMethodField()
+    avatar = serializers.ImageField(required=False, allow_null=True) # 上传图片
+    class Meta:
+        model = User
+        fields = [
+            'id',
+            'username',
+            'full_name',
+            'avatar',
+        ]
+
+    def get_full_name(self, odj) -> str:
+        return f'{odj.first_name} {odj.last_name}'
+
+
+class UserDetailSerializer(serializers.ModelSerializer):
+    full_name = serializers.SerializerMethodField()
+    avatar = serializers.ImageField(required=False, allow_null=True)
+    class Meta:
+        model = User
+        fields = [
+            'id',
+            'username',
+            'full_name',
+            'avatar',
+            'gender',
+            'is_verified',
+            'rating'
+        ]
+
+    def get_full_name(self, odj) -> str:
+        return f'{odj.first_name} {odj.last_name}'
```
U have two matching serializers. The only difference between them is more extended list of fields in the second case.

Try creating the `UserBaseSerializer` class and inheriting the `UserSerializer` and `UserDetailSerializer` classes with it.

ArtyomYapryntsev reviewed (changes requested):
Very good, try to fix the issues

HerdaSinama committed:
Fix serializers

now UserDetailSerializer expand UserSerializer

HerdaSinama committed:
blyat

HerdaSinama committed:
Fix model method and some typo

- validator_avatar_weight: fix error message.
- is_creator: was deleted.
- remove_member, set_status_active, set_status_disbanded:
method is_creator is not used, now the role is passed to the method.

HerdaSinama committed:
Add docs for models

HerdaSinama committed:
Add new views according to task

- UserListView: return list of users, then returns all users whose username starts with the '...'
- UserDetailView: depending on the request GET or PUT user information
- AnotherUserView:
return the specific user data

HerdaSinama commented on server/apps/acсount/views.py:
```diff
@@ -0,0 +1,49 @@
+    permission_classes = (IsAuthenticated,)
+
+    def get_queryset(self):
+        queryset = User.objects.all()
```
does it make good practice if instead .all() to use .only(ser. field)?

ArtyomYapryntsev replied:
U can ask about it in the chatgpt :smile: 

In my practice, I have always used a simple `.all()` construction and filtered instances using serializer fields.

I think that in development, using `.all()` can help solve some problems.
But in production mode, it is better to use `.only(...fields)` to extract only the necessary fields from the database (this way the size of the database response will be smaller, which means that the request execution time will decrease).

HerdaSinama committed:
Update class UserDetailView (like optimus prime)

- UserDetailView: now contains AnotherUserView
- AnotherUserView: was deleted

HerdaSinama committed:
Add docs for views

HerdaSinama committed:
Init tests modul
